src/kademlia/sample_infohashes.cpp

```diff
@@ -0,0 +1,144 @@
+/*
+
+Copyright (c) 2017, Arvid Norberg, Alden Torres
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in
+      the documentation and/or other materials provided with the distribution.
+    * Neither the name of the author nor the names of its
+      contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#include <libtorrent/kademlia/sample_infohashes.hpp>
+#include <libtorrent/kademlia/dht_observer.hpp>
+#include <libtorrent/kademlia/node.hpp>
+#include <libtorrent/kademlia/io.hpp>
+#include <libtorrent/performance_counters.hpp>
+#include <libtorrent/aux_/numeric_cast.hpp>
+#include <libtorrent/socket_io.hpp>
+
+namespace libtorrent { namespace dht
+{
+
+sample_infohashes::sample_infohashes(node& dht_node
+	, node_id const& target
+	, data_callback const& dcallback)
+	: traversal_algorithm(dht_node, target)
+	, m_data_callback(dcallback) {}
+
+char const* sample_infohashes::name() const { return "sample_infohashes"; }
+
+void sample_infohashes::got_samples(time_duration interval
+	, int num, std::vector<sha1_hash> samples
+	, std::vector<std::pair<sha1_hash, udp::endpoint>> nodes)
+{
+	if (m_data_callback)
```

Comment:
is there ever any use case where this callback may not be set? If so, would it make sense to even issue the request?